Javalin Raspberry Pi Tutorial (#82)
javalin/javalin.github.io: _posts/tutorials/2020-09-05-javalin-raspberry-pi-example.md

---
layout: tutorial
title: "Using Javalin with Kotlin to deploy a simple REST API to a Raspberry Pi"
author: <a href="https://github.com/NickM-27" target="_blank">Nick Mowen</a>

> date: 9/5/2020

Reviewer: ```suggestion
date: 2020-09-05
```

summarytitle: Deploy Kotlin REST API to Raspberry Pi.
summary: Use Kotlin with Javalin to deploy a REST API to run on a Raspberry Pi"
language: Kotlin
---

## What You Will Learn

* Setting up Kotlin with Gradle
* Creating a simple REST API using Javalin

> * Deploying and running the API on an embedded Jetty runner on a Raspberry Pi

Reviewer: ```suggestion
* Deploying and running the API on an embedded Jetty running on a Raspberry Pi
```

The instructions for this tutorial will be used with IntelliJ IDEA. It is recommended to download the free community edition of IDEA.

## Setting up Kotlin with Gradle (in IntelliJ IDEA)

* `File` `->` `New` `->` `Project`
* `Gradle` `->` `Kotlin/JVM`
* Set the project name and other attributes of the project
* Create `src/main/kotlin/com/org/example/Main.kt`

Kotlin offers a `main` with and without args, in our case we don't need args so we can define main as:

``` kotlin
fun main() {

}
```

## Setting Up Javalin Dependencies

> First, you will need to change the kotlin dependency from `implementation` to `compile` which will be necessary later for building the runnable jar.

Reviewer: ```suggestion
First, you will need to change the Kotlin dependency from `implementation` to `compile` which will be necessary later for building the runnable jar.
```

You will need to add `jcenter` to the respositories that dependencies are included in:

``` groovy
repositories {
    mavenCentral()
    jcenter()
}
```

Then, add these dependencies as well:

``` groovy
compile 'io.javalin:javalin:3.10.1'

> compile 'com.fasterxml.jackson.core:jackson-databind:2.10.3' // Necessary to be included for running on Raspberry Pi

Reviewer: ```suggestion
compile 'com.fasterxml.jackson.core:jackson-databind:2.10.3' // Necessary for serializing JSON
```

> compile 'com.fasterxml.jackson.module:jackson-module-kotlin:2.10.3' // Necessary to be included for running on Raspberry Pi

Reviewer: ```suggestion
compile 'com.fasterxml.jackson.module:jackson-module-kotlin:2.10.3' // Necessary for serializing JSON 
```

> compile 'org.slf4j:slf4j-simple:1.7.30' // Necessary to be included for running on Raspberry Pi

Reviewer: ```suggestion
compile 'org.slf4j:slf4j-simple:1.7.30' // Necessary to view logging output
```

```

## Setting Up Javalin

We need to create the Javalin app with the IP address and port to be used on the Raspberry Pi:

``` kotlin

> fun main() {
>     val app = Javalin.create().apply {
>         exception(Exception::class.java) { e, _ -> e.printStackTrace() }
>     }.start("your.raspberry.pi.ipaddress", 8080)
> }

Reviewer: ```suggestion
fun main() {
    val ipAddress = "0.0.0.0" // change this to your Raspberry Pi's IP address
    val app = Javalin.create().apply {
        exception(Exception::class.java) { e, _ -> e.printStackTrace() }
    }.start(ipAddress, 8080)
}
```

    
    fun addStringData(data: StringData) {
        strings.add(data)
    }
    
    fun removeStringData(stringId: Long): Boolean = strings.removeIf { it.id == stringId }
}
```

## Creating Simple REST API To Test With

Now we can add the REST API with our app in main:

``` kotlin
fun main() {
    val stringDao = StringDao()

    val app = Javalin.create().apply {
        exception(Exception::class.java) { e, _ -> e.printStackTrace() }
    }.start("your.raspberry.pi.ipaddress", 8080)

    app.routes {
        get("/strings") { context ->
            context.json(data.strings)
        }

        post("/strings") { context ->
            val stringData = context.body<StringData>()
            stringDao.addStringData(stringData)
            context.status(201)
        }

        delete("/strings/:stringId") { context ->
            stringDao.removeStringData(context.pathParam("stringId").toLongOrNull() ?: -1).let {
                if (it)
                    context.status(204)
                else
                    context.status(400)
            }
        }
    }
}
```

## Building A Fat Jar

> A fat jar is a runnable java file (like a normal jar) which contains all its necessary dependencies within itself to ensure it can run.

Reviewer: ```suggestion
A fat jar is a single file which contains all the compiled Java classes and dependencies required to run an application.
```
